File: backend/src/games/carcassonne/features.py
"""Feature tracking: creation, merging, completion detection."""

from __future__ import annotations

from uuid import uuid4

from src.engine.models import Event
from src.games.carcassonne.tiles import TILE_LOOKUP, get_rotated_features
from src.games.carcassonne.types import (
    DIRECTIONS,
    OPPOSITE_DIRECTION,
    Feature,
    FeatureType,
    Position,
    rotate_edges,
)


def initialize_features_from_tile(
    tile_type_id: str,
    position_key: str,
    rotation: int,
) -> tuple[dict[str, dict], dict[str, dict[str, str]]]:
    """Create features for a single tile (used for the starting tile).

    Returns:
        features: {feature_id: Feature-as-dict}
        tile_feature_map: {position_key: {meeple_spot: feature_id}}
    """
    features: dict[str, dict] = {}
    tile_feature_map: dict[str, dict[str, str]] = {position_key: {}}

    rotated_features = get_rotated_features(tile_type_id, rotation)
    tile_def = TILE_LOOKUP[tile_type_id]
    rotated_edges_map = rotate_edges(tile_def.edges, rotation)

    for tile_feat in rotated_features:
        feature_id = str(uuid4())

        # Calculate open edges for this feature
        open_edges: list[list[str]] = []
        for edge_dir in tile_feat.edges:
            # This edge is open (no adjacent tile connected yet)
            open_edges.append([position_key, edge_dir])

        # Count pennants
        pennants = 1 if tile_feat.has_pennant else 0

        feature = Feature(
            feature_id=feature_id,
            feature_type=tile_feat.feature_type,
            tiles=[position_key],
            meeples=[],
            is_complete=False,
            pennants=pennants,
            open_edges=open_edges,
        )
        features[feature_id] = feature.model_dump()

        # Map meeple spots to this feature
        for spot in tile_feat.meeple_spots:
            tile_feature_map[position_key][spot] = feature_id

    return features, tile_feature_map


def create_and_merge_features(
    game_data: dict,
    tile_type_id: str,
    position_key: str,
    rotation: int,
) -> list[Event]:
    """Place a tile's features on the board and merge with adjacent features.

    Modifies game_data["features"] and game_data["tile_feature_map"] in place.
    Returns a list of merge events.
    """
    events: list[Event] = []
    features = game_data["features"]
    tile_feature_map = game_data["tile_feature_map"]
    board_tiles = game_data["board"]["tiles"]

    # Initialize the feature map for this tile position
    tile_feature_map[position_key] = {}

    rotated_features = get_rotated_features(tile_type_id, rotation)
    tile_def = TILE_LOOKUP[tile_type_id]
    rotated_edges_map = rotate_edges(tile_def.edges, rotation)

    # Step 1: Create features for the new tile
    new_feature_ids: list[str] = []
    # Map from edge direction to the feature that touches that edge on this new tile
    edge_to_feature: dict[str, str] = {}

    for tile_feat in rotated_features:
        feature_id = str(uuid4())

        open_edges: list[list[str]] = []
        for edge_dir in tile_feat.edges:
            open_edges.append([position_key, edge_dir])

        pennants = 1 if tile_feat.has_pennant else 0

        feature = Feature(
            feature_id=feature_id,
            feature_type=tile_feat.feature_type,
            tiles=[position_key],
            meeples=[],
            is_complete=False,
            pennants=pennants,
            open_edges=open_edges,
        )
        features[feature_id] = feature.model_dump()
        new_feature_ids.append(feature_id)

        for spot in tile_feat.meeple_spots:
            tile_feature_map[position_key][spot] = feature_id

        for edge_dir in tile_feat.edges:
            edge_to_feature[edge_dir] = feature_id

    # Step 2: Merge with adjacent tiles
    pos = Position.from_key(position_key)

    for direction in DIRECTIONS:
        neighbor_pos = pos.neighbor(direction)
        neighbor_key = neighbor_pos.to_key()

        if neighbor_key not in board_tiles:
            continue

        if direction not in edge_to_feature:
            # No feature on this edge (shouldn't happen for non-field edges,
            # but fields may not explicitly list all edges)
            continue

        our_feature_id = edge_to_feature[direction]
        # Resolve to current feature (may have been merged already)
        our_feature_id = _resolve_feature_id(features, our_feature_id, tile_feature_map)

        opposite_dir = OPPOSITE_DIRECTION[direction]

        # Find the feature on the adjacent tile that touches the opposite edge
        adj_feature_id = _find_feature_at_edge(
            tile_feature_map, features, neighbor_key, opposite_dir
        )

        if adj_feature_id is None:
            continue

        adj_feature_id = _resolve_feature_id(features, adj_feature_id, tile_feature_map)

        if our_feature_id == adj_feature_id:
            # Already the same feature (can happen after previous merges)
            # Still need to remove the connecting open edges
            _remove_open_edge(features[our_feature_id], position_key, direction)
            _remove_open_edge(features[our_feature_id], neighbor_key, opposite_dir)
            continue

        # Verify same feature type
        if features[our_feature_id]["feature_type"] != features[adj_feature_id]["feature_type"]:
            continue

        # Merge: absorb adj into ours
        merged_id = _merge_features(
            features, tile_feature_map, our_feature_id, adj_feature_id
        )

        # Remove the connecting open edges
        _remove_open_edge(features[merged_id], position_key, direction)
        _remove_open_edge(features[merged_id], neighbor_key, opposite_dir)

        # Update edge_to_feature for subsequent merges
        for d, fid in edge_to_feature.items():
            if fid == adj_feature_id or fid == our_feature_id:
                edge_to_feature[d] = merged_id

        events.append(Event(
            event_type="feature_merged",
            payload={
                "surviving_feature": merged_id,
                "merged_feature": adj_feature_id if merged_id == our_feature_id else our_feature_id,
                "feature_type": features[merged_id]["feature_type"],
            },
        ))

    return events


def _resolve_feature_id(
    features: dict[str, dict],
    feature_id: str,
    tile_feature_map: dict[str, dict[str, str]],
) -> str:
    """Resolve a feature ID that may have been merged into another."""
    if feature_id in features:
        return feature_id
    # Search for which feature absorbed this one
    for fid, feat in features.items():
        if feature_id in feat.get("_merged_from", []):
            return fid
    return feature_id


def _find_feature_at_edge(
    tile_feature_map: dict[str, dict[str, str]],
    features: dict[str, dict],
    position_key: str,
    direction: str,
) -> str | None:
    """Find the feature ID that touches a specific edge on a tile."""
    if position_key not in tile_feature_map:
        return None

    for spot, fid in tile_feature_map[position_key].items():
        if fid not in features:
            continue
        feat = features[fid]
        # Check if this feature has an open edge at this position+direction
        for oe in feat.get("open_edges", []):
            if oe[0] == position_key and oe[1] == direction:
                return fid
        # Also check if the feature's tiles include this position and the
        # original tile feature touched this edge
        # We check open_edges which should contain all un-connected edges
    return None


def _merge_features(
    features: dict[str, dict],
    tile_feature_map: dict[str, dict[str, str]],
    feature_a_id: str,
    feature_b_id: str,
) -> str:
    """Merge feature_b into feature_a. Returns the surviving feature ID."""
    a = features[feature_a_id]
    b = features[feature_b_id]

    # Combine tiles (deduplicate)
    combined_tiles = list(set(a["tiles"] + b["tiles"]))
    a["tiles"] = combined_tiles

    # Combine meeples
    a["meeples"] = a.get("meeples", []) + b.get("meeples", [])

    # Combine pennants
    a["pennants"] = a.get("pennants", 0) + b.get("pennants", 0)

    # Combine open edges
    a["open_edges"] = a.get("open_edges", []) + b.get("open_edges", [])

    # Track merged IDs for resolution
    merged_from = a.get("_merged_from", [])
    merged_from.append(feature_b_id)
    merged_from.extend(b.get("_merged_from", []))
    a["_merged_from"] = merged_from

    # Update tile_feature_map: all references to feature_b now point to feature_a
    for pos_key, spots in tile_feature_map.items():
        for spot, fid in spots.items():
            if fid == feature_b_id:
                spots[spot] = feature_a_id

    # Remove feature_b
    del features[feature_b_id]

    return feature_a_id


def _remove_open_edge(feature_dict: dict, position_key: str, direction: str) -> None:
    """Remove a specific open edge from a feature."""
    open_edges = feature_dict.get("open_edges", [])
    feature_dict["open_edges"] = [
        oe for oe in open_edges
        if not (oe[0] == position_key and oe[1] == direction)
    ]


def is_feature_complete(game_data: dict, feature: dict) -> bool:
    """Check if a feature is complete.

    - City/Road: complete when no open edges remain
    - Monastery: complete when all 8 surrounding tiles are placed
    - Field: never complete during the game
    """
    ft = feature["feature_type"]

    if ft == FeatureType.FIELD or ft == "field":
        return False

    if ft == FeatureType.MONASTERY or ft == "monastery":
        if not feature["tiles"]:
            return False
        pos = Position.from_key(feature["tiles"][0])
        board_tiles = game_data["board"]["tiles"]
        for surrounding in pos.all_surrounding():
            if surrounding.to_key() not in board_tiles:
                return False
        return True

    # City or Road: complete when no open edges
    return len(feature.get("open_edges", [])) == 0


def check_monastery_completion(
    game_data: dict, position_key: str
) -> tuple[list[Event], dict[str, int]]:
    """Check if any monasteries near the placed tile are now complete.

    When a tile is placed, it might complete monasteries on adjacent tiles
    (including diagonals) or on itself.

    Returns:
        events: scoring events
        scores: {player_id: points}
    """
    from src.games.carcassonne.scoring import score_completed_feature

    events: list[Event] = []
    scores: dict[str, int] = {}
    features = game_data["features"]
    board_tiles = game_data["board"]["tiles"]
    tile_feature_map = game_data["tile_feature_map"]

    pos = Position.from_key(position_key)

    # Check the placed tile and all 8 surrounding tiles for monasteries
    positions_to_check = [pos] + pos.all_surrounding()

    for check_pos in positions_to_check:
        check_key = check_pos.to_key()
        if check_key not in tile_feature_map:
            continue

        for spot, feature_id in tile_feature_map[check_key].items():
            if feature_id not in features:
                continue
            feature = features[feature_id]
            if feature.get("feature_type") not in (FeatureType.MONASTERY, "monastery"):
                continue
            if feature.get("is_complete"):
                continue

            if is_feature_complete(game_data, feature):
                feature["is_complete"] = True
                point_awards = score_completed_feature(feature)

                for pid, points in point_awards.items():
                    scores[pid] = scores.get(pid, 0) + points
                    events.append(Event(
                        event_type="feature_scored",
                        player_id=pid,
                        payload={
                            "feature_id": feature_id,
                            "feature_type": "monastery",
                            "points": points,
                            "tiles": feature["tiles"],
                        },
                    ))

                # Return meeples
                from src.games.carcassonne.meeples import return_meeples
                meeple_events = return_meeples(game_data, feature)
                events.extend(meeple_events)

    return events, scores

File: game-engine/src/games/carcassonne/features.rs
//! Feature tracking: creation, merging, and completion detection.
//! Mirrors backend/src/games/carcassonne/features.py.

use std::collections::HashMap;

use uuid::Uuid;

use crate::engine::models::Event;
use super::tiles::get_rotated_features;
use super::types::*;

/// Get the opposite edge identifier. Handles compound edges like "E:N" → "W:N".
fn opposite_edge(edge: &str) -> String {
    if let Some((direction, side)) = edge.split_once(':') {
        format!("{}:{}", opposite_direction(direction), side)
    } else {
        opposite_direction(edge).to_string()
    }
}

/// Get the cardinal direction from an edge (simple or compound).
fn edge_direction(edge: &str) -> &str {
    edge.split(':').next().unwrap()
}

/// Create features for a single tile (used for the starting tile).
pub fn initialize_features_from_tile(
    tile_type_id: &str,
    position_key: &str,
    rotation: u32,
) -> (HashMap<String, serde_json::Value>, HashMap<String, HashMap<String, String>>) {
    let mut features: HashMap<String, serde_json::Value> = HashMap::new();
    let mut tile_feature_map: HashMap<String, HashMap<String, String>> = HashMap::new();
    tile_feature_map.insert(position_key.to_string(), HashMap::new());

    let rotated_features = get_rotated_features(tile_type_id, rotation);

    for tile_feat in &rotated_features {
        let feature_id = Uuid::new_v4().to_string();

        let open_edges: Vec<[String; 2]> = tile_feat
            .edges
            .iter()
            .map(|e| [position_key.to_string(), e.clone()])
            .collect();

        let pennants: u32 = if tile_feat.has_pennant { 1 } else { 0 };

        let feature = Feature {
            feature_id: feature_id.clone(),
            feature_type: tile_feat.feature_type,
            tiles: vec![position_key.to_string()],
            meeples: vec![],
            is_complete: false,
            pennants,
            open_edges,
        };
        features.insert(feature_id.clone(), serde_json::to_value(&feature).unwrap());

        let spots = tile_feature_map.get_mut(position_key).unwrap();
        for spot in &tile_feat.meeple_spots {
            spots.insert(spot.clone(), feature_id.clone());
        }
    }

    (features, tile_feature_map)
}

/// Place a tile's features on the board and merge with adjacent features.
/// Modifies game_data in place. Returns merge events.
pub fn create_and_merge_features(
    game_data: &mut serde_json::Value,
    tile_type_id: &str,
    position_key: &str,
    rotation: u32,
) -> Vec<Event> {
    let mut events = Vec::new();

    let rotated_features = get_rotated_features(tile_type_id, rotation);

    // Initialize the feature map for this tile position
    game_data["tile_feature_map"][position_key] = serde_json::json!({});

    // Map from edge direction to the feature that touches that edge
    let mut edge_to_feature: Vec<(String, String)> = Vec::new();
    let mut new_feature_ids: Vec<String> = Vec::new();

    // Step 1: Create features for the new tile
    for tile_feat in &rotated_features {
        let feature_id = Uuid::new_v4().to_string();

        let open_edges: Vec<serde_json::Value> = tile_feat
            .edges
            .iter()
            .map(|e| serde_json::json!([position_key, e]))
            .collect();

        let pennants: u32 = if tile_feat.has_pennant { 1 } else { 0 };

        let feature = serde_json::json!({
            "feature_id": feature_id,
            "feature_type": tile_feat.feature_type,
            "tiles": [position_key],
            "meeples": [],
            "is_complete": false,
            "pennants": pennants,
            "open_edges": open_edges,
        });

        game_data["features"][&feature_id] = feature;
        new_feature_ids.push(feature_id.clone());

        for spot in &tile_feat.meeple_spots {
            game_data["tile_feature_map"][position_key][spot.as_str()] =
                serde_json::json!(feature_id);
        }

        for edge_dir in &tile_feat.edges {
            edge_to_feature.push((edge_dir.clone(), feature_id.clone()));
        }
    }

    // Step 2: Merge with adjacent tiles
    let pos = Position::from_key(position_key);

    for i in 0..edge_to_feature.len() {
        let (edge_key, mut our_feature_id) = edge_to_feature[i].clone();
        let direction = edge_direction(&edge_key);
        let neighbor_pos = pos.neighbor(direction);
        let neighbor_key = neighbor_pos.to_key();

        // Check if neighbor tile exists
        if game_data["board"]["tiles"].get(&neighbor_key).is_none() {
            continue;
        }

        // Resolve to current feature (may have been merged)
        our_feature_id = resolve_feature_id(game_data, &our_feature_id);

        let opp_edge = opposite_edge(&edge_key);

        // Find adjacent feature
        let adj_feature_id = match find_feature_at_edge(game_data, &neighbor_key, &opp_edge) {
            Some(id) => resolve_feature_id(game_data, &id),
            None => continue,
        };

        if our_feature_id == adj_feature_id {
            // Already same feature — just remove connecting open edges
            remove_open_edge(game_data, &our_feature_id, position_key, &edge_key);
            remove_open_edge(game_data, &our_feature_id, &neighbor_key, &opp_edge);
            continue;
        }

        // Verify same feature type
        let our_type = game_data["features"][&our_feature_id]["feature_type"].as_str().unwrap_or("");
        let adj_type = game_data["features"][&adj_feature_id]["feature_type"].as_str().unwrap_or("");
        if our_type != adj_type {
            continue;
        }

        // Merge: absorb adj into ours
        let merged_id = merge_features(game_data, &our_feature_id, &adj_feature_id);

        remove_open_edge(game_data, &merged_id, position_key, &edge_key);
        remove_open_edge(game_data, &merged_id, &neighbor_key, &opp_edge);

        // Update edge_to_feature for subsequent merges
        for (_, fid) in edge_to_feature.iter_mut() {
            if *fid == adj_feature_id || *fid == our_feature_id {
                *fid = merged_id.clone();
            }
        }

        let merged_type = game_data["features"][&merged_id]["feature_type"].clone();
        events.push(Event {
            event_type: "feature_merged".to_string(),
            player_id: None,
            payload: serde_json::json!({
                "surviving_feature": merged_id,
                "merged_feature": if merged_id == our_feature_id { &adj_feature_id } else { &our_feature_id },
                "feature_type": merged_type,
            }),
        });
    }

    events
}

/// Resolve a feature ID that may have been merged into another.
fn resolve_feature_id(game_data: &serde_json::Value, feature_id: &str) -> String {
    if game_data["features"].get(feature_id).is_some() {
        return feature_id.to_string();
    }
    // Search for which feature absorbed this one
    if let Some(features) = game_data["features"].as_object() {
        for (fid, feat) in features {
            if let Some(merged_from) = feat.get("_merged_from") {
                if let Some(arr) = merged_from.as_array() {
                    for v in arr {
                        if v.as_str() == Some(feature_id) {
                            return fid.clone();
                        }
                    }
                }
            }
        }
    }
    feature_id.to_string()
}

/// Find the feature ID that touches a specific edge on a tile.
fn find_feature_at_edge(
    game_data: &serde_json::Value,
    position_key: &str,
    direction: &str,
) -> Option<String> {
    let tfm = game_data["tile_feature_map"].get(position_key)?;
    let spots = tfm.as_object()?;

    for (_spot, fid_val) in spots {
        let fid = fid_val.as_str()?;
        let feat = game_data["features"].get(fid)?;
        if let Some(open_edges) = feat.get("open_edges") {
            if let Some(arr) = open_edges.as_array() {
                for oe in arr {
                    if let Some(oe_arr) = oe.as_array() {
                        if oe_arr.len() == 2
                            && oe_arr[0].as_str() == Some(position_key)
                            && oe_arr[1].as_str() == Some(direction)
                        {
                            return Some(fid.to_string());
                        }
                    }
                }
            }
        }
    }
    None
}

/// Merge feature_b into feature_a. Returns the surviving feature ID.
fn merge_features(
    game_data: &mut serde_json::Value,
    feature_a_id: &str,
    feature_b_id: &str,
) -> String {
    // Read feature_b data before removing it
    let b = game_data["features"][feature_b_id].clone();

    let a = &mut game_data["features"][feature_a_id];

    // Combine tiles (deduplicate)
    let mut tiles: Vec<String> = Vec::new();
    if let Some(arr) = a["tiles"].as_array() {
        for v in arr { if let Some(s) = v.as_str() { tiles.push(s.to_string()); } }
    }
    if let Some(arr) = b["tiles"].as_array() {
        for v in arr {
            if let Some(s) = v.as_str() {
                if !tiles.contains(&s.to_string()) {
                    tiles.push(s.to_string());
                }
            }
        }
    }
    a["tiles"] = serde_json::json!(tiles);

    // Combine meeples
    let mut meeples: Vec<serde_json::Value> = Vec::new();
    if let Some(arr) = a["meeples"].as_array() { meeples.extend(arr.iter().cloned()); }
    if let Some(arr) = b["meeples"].as_array() { meeples.extend(arr.iter().cloned()); }
    a["meeples"] = serde_json::json!(meeples);

    // Combine pennants
    let pa = a["pennants"].as_u64().unwrap_or(0);
    let pb = b["pennants"].as_u64().unwrap_or(0);
    a["pennants"] = serde_json::json!(pa + pb);

    // Combine open edges
    let mut open_edges: Vec<serde_json::Value> = Vec::new();
    if let Some(arr) = a["open_edges"].as_array() { open_edges.extend(arr.iter().cloned()); }
    if let Some(arr) = b["open_edges"].as_array() { open_edges.extend(arr.iter().cloned()); }
    a["open_edges"] = serde_json::json!(open_edges);

    // Track merged IDs
    let mut merged_from: Vec<String> = Vec::new();
    if let Some(arr) = a.get("_merged_from").and_then(|v| v.as_array()) {
        for v in arr { if let Some(s) = v.as_str() { merged_from.push(s.to_string()); } }
    }
    merged_from.push(feature_b_id.to_string());
    if let Some(arr) = b.get("_merged_from").and_then(|v| v.as_array()) {
        for v in arr { if let Some(s) = v.as_str() { merged_from.push(s.to_string()); } }
    }
    a["_merged_from"] = serde_json::json!(merged_from);

    // Update tile_feature_map: all references to feature_b → feature_a
    if let Some(tfm) = game_data["tile_feature_map"].as_object_mut() {
        for (_pos, spots) in tfm.iter_mut() {
            if let Some(spots_obj) = spots.as_object_mut() {
                for (_spot, fid) in spots_obj.iter_mut() {
                    if fid.as_str() == Some(feature_b_id) {
                        *fid = serde_json::json!(feature_a_id);
                    }
                }
            }
        }
    }

    // Remove feature_b
    if let Some(features) = game_data["features"].as_object_mut() {
        features.remove(feature_b_id);
    }

    feature_a_id.to_string()
}

/// Remove a specific open edge from a feature.
fn remove_open_edge(
    game_data: &mut serde_json::Value,
    feature_id: &str,
    position_key: &str,
    direction: &str,
) {
    if let Some(feat) = game_data["features"].get_mut(feature_id) {
        if let Some(open_edges) = feat.get_mut("open_edges") {
            if let Some(arr) = open_edges.as_array() {
                let filtered: Vec<serde_json::Value> = arr
                    .iter()
                    .filter(|oe| {
                        if let Some(oe_arr) = oe.as_array() {
                            !(oe_arr.len() == 2
                                && oe_arr[0].as_str() == Some(position_key)
                                && oe_arr[1].as_str() == Some(direction))
                        } else {
                            true
                        }
                    })
                    .cloned()
                    .collect();
                *open_edges = serde_json::json!(filtered);
            }
        }
    }
}

/// Check if a feature is complete.
pub fn is_feature_complete(game_data: &serde_json::Value, feature: &serde_json::Value) -> bool {
    let ft = feature["feature_type"].as_str().unwrap_or("");

    if ft == "field" {
        return false;
    }

    if ft == "monastery" {
        let tiles = match feature["tiles"].as_array() {
            Some(arr) => arr,
            None => return false,
        };
        if tiles.is_empty() {
            return false;
        }
        let pos = Position::from_key(tiles[0].as_str().unwrap());
        let board_tiles = &game_data["board"]["tiles"];
        for surrounding in pos.all_surrounding() {
            if board_tiles.get(&surrounding.to_key()).is_none() {
                return false;
            }
        }
        return true;
    }

    // City or Road: complete when no open edges
    match feature.get("open_edges").and_then(|v| v.as_array()) {
        Some(arr) => arr.is_empty(),
        None => true,
    }
}

/// Check if any monasteries near the placed tile are now complete.
pub fn check_monastery_completion(
    game_data: &mut serde_json::Value,
    position_key: &str,
) -> (Vec<Event>, HashMap<String, i64>) {
    use super::scoring::score_completed_feature;
    use super::meeples::return_meeples;

    let mut events = Vec::new();
    let mut scores: HashMap<String, i64> = HashMap::new();

    let pos = Position::from_key(position_key);
    let mut positions_to_check = vec![pos];
    positions_to_check.extend(pos.all_surrounding());

    for check_pos in positions_to_check {
        let check_key = check_pos.to_key();

        // Collect monastery feature IDs at this position
        let monastery_ids: Vec<String> = {
            let tfm = &game_data["tile_feature_map"];
            let Some(spots) = tfm.get(&check_key).and_then(|v| v.as_object()) else {
                continue;
            };
            spots
                .values()
                .filter_map(|fid_val| {
                    let fid = fid_val.as_str()?;
                    let feat = game_data["features"].get(fid)?;
                    if feat["feature_type"].as_str() == Some("monastery")
                        && !feat["is_complete"].as_bool().unwrap_or(false)
                    {
                        Some(fid.to_string())
                    } else {
                        None
                    }
                })
                .collect()
        };

        for feature_id in monastery_ids {
            let feature = &game_data["features"][&feature_id];
            if !is_feature_complete(game_data, feature) {
                continue;
            }

            // Mark complete
            game_data["features"][&feature_id]["is_complete"] = serde_json::json!(true);

            let feature = &game_data["features"][&feature_id];
            let point_awards = score_completed_feature(feature);

            for (pid, points) in &point_awards {
                *scores.entry(pid.clone()).or_insert(0) += points;
                events.push(Event {
                    event_type: "feature_scored".to_string(),
                    player_id: Some(pid.clone()),
                    payload: serde_json::json!({
                        "feature_id": feature_id,
                        "feature_type": "monastery",
                        "points": points,
                        "tiles": game_data["features"][&feature_id]["tiles"],
                    }),
                });
            }

            // Return meeples
            let meeple_events = return_meeples(game_data, &feature_id);
            events.extend(meeple_events);
        }
    }

    (events, scores)
}
